arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,294 @@
+#if defined(CONFIG_MPU_STACK_GUARD)
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM \
+	(CONFIG_MAX_DOMAIN_PARTITIONS + 2)
 #else
-	u32_t guard_start = thread->stack_info.start;
-#endif
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM \
+	(CONFIG_MAX_DOMAIN_PARTITIONS + 1)
+#endif /* CONFIG_MPU_STACK_GUARD */
+#else /* CONFIG_USERSPACE */
+#if defined(CONFIG_MPU_STACK_GUARD)
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM 1
+#else
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM 0
+#endif /* CONFIG_MPU_STACK_GUARD */
+#endif /* CONFIG_USERSPACE */
 
-	arm_core_mpu_disable();
-	arm_core_mpu_configure(THREAD_STACK_GUARD_REGION, guard_start,
-			       guard_size);
-	arm_core_mpu_enable();
-}
-#endif
+/* Convenience macros to denote the start address and the size of the system
+ * memory area, where dynamic memory regions may be programmed at run-time.
+ */
+#if defined(CONFIG_APP_SHARED_MEM)
+#define _MPU_DYNAMIC_REGIONS_AREA_START ((u32_t)&_app_smem_start)
+#else
+#define _MPU_DYNAMIC_REGIONS_AREA_START ((u32_t)&__kernel_ram_start)
+#endif /* CONFIG_APP_SHARED_MEM */
+#define _MPU_DYNAMIC_REGIONS_AREA_SIZE ((u32_t)&__kernel_ram_end - \
+		_MPU_DYNAMIC_REGIONS_AREA_START)
 
-#if defined(CONFIG_USERSPACE)
-/*
- * @brief Configure MPU user context
+/**
+ * @brief Use the HW-specific MPU driver to program
+ *        the static MPU regions.
+ *
+ * Program the static MPU regions using the HW-specific MPU driver. The
+ * function is meant to be invoked only once upon system initialization.
  *
- * This function configures the thread's user context.
- * The functionality is meant to be used during context switch.
+ * If the function attempts to configure a number of regions beyond the
+ * MPU HW limitations, the system behavior will be undefined.
  *
- * @param thread thread info data structure.
+ * For some MPU architectures, such as the unmodified ARMv8-M MPU,
+ * the function must execute with MPU enabled.
  */
-void configure_mpu_user_context(struct k_thread *thread)
+void _arch_configure_static_mpu_regions(void)
 {
-	LOG_DBG("configure user thread %p's context", thread);
-	arm_core_mpu_disable();
-	arm_core_mpu_configure_user_context(thread);
-	arm_core_mpu_enable();
+	/* Define a constant array of k_mem_partition objects
+	 * to hold the configuration of the respective static
+	 * MPU regions.
+	 */
+	const struct k_mem_partition static_regions[] = {
+#if defined(CONFIG_APPLICATION_MEMORY)
+		{
+		.start = (u32_t)&__app_ram_start,
+		.size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start,
+		.attr = K_MEM_PARTITION_P_RW_U_RW,
+		},
+#endif /* CONFIG_APPLICATION_MEMORY */
+#if defined(CONFIG_COVERAGE_GCOV) && defined(CONFIG_USERSPACE)
+		{
+		.start = (u32_t)&__gcov_bss_start,
+		.size = (u32_t)&__gcov_bss_end - (u32_t)&__gcov_bss_start,
+		.attr = K_MEM_PARTITION_P_RW_U_RW,
+		},
+#endif /* CONFIG_COVERAGE_GCOV && CONFIG_USERSPACE */
+#if defined(CONFIG_NOCACHE_MEMORY)
+		{
+		.start = (u32_t)&_nocache_ram_start,
+		.size = (u32_t)&_nocache_ram_end - (u32_t)&_nocache_ram_start,
```

Comment:
> _start/_end symbols already have char type in linker-defs.h. I don't see why these casts are needed, 

You have to cast because although they are char type, they have no interesting value, it's their *memory address* that is what we want, and that is a pointer, hence the use of &

Reply:
I suppose, addressed?

Reply:
To be clear: the result of subtracting two char* values has type size_t and is in units of bytes, which is AFAICT exactly what you want.  So `(&_foo_end - &foo_start)` should give you an integral value you can use without casting.  But not a big deal.